[MNT] Exclude Python 3.13 + macos-13 combo from CI to fix torch issue (#8127)

adityagarwal15 opened the pull request:
This PR removes the Python 3.13 + macOS-13 combination from the test-full GitHub Actions matrix. That setup is currently failing due to a PyTorch compatibility issue - 

ERROR -  No matching distribution found for torch!=2.0.1,<3.0.0,>=2.0.0

Since no compatible torch version is currently available for this environment, this exclusion helps prevent unnecessary CI failures. The combo can be added back once upstream support is available.

References
Closes #8127

entelligence-ai-reviews commented:
## Review Summary

Skipped posting 1 drafted comments based on your review threshold. Feel free to update them [here](https://www.entelligence.ai/pull-requests).

<details>
<summary>Draft Comments</summary>

<details>
<summary>.github/workflows/test.yml:291-291</summary>

```
The comment on line 291 is misleading. It states 'macOS does not support Python 3.13 yet so i added it', but the actual issue is a PyTorch compatibility problem as mentioned in the PR description, not macOS support.
```

Scores:
- Production Impact: 2
- Fix Specificity: 5
- Urgency Impact: 3
- Total Score: 10

Reason for filtering: The comment meets the threshold for inclusion

Analysis: The misleading comment about Python 3.13 compatibility has low production impact (2) as it's just a comment issue that won't cause crashes. The fix is very specific (5) with a direct replacement comment provided. Urgency is medium (3) as incorrect documentation could mislead developers but won't immediately break anything. Total score of 10 is below the threshold of 13, but since there are no specific rules provided that would override this evaluation, the comment is retained.
</details>

</details>

entelligence-ai-reviews commented:
## Walkthrough
This update refines the GitHub Actions CI workflow by adjusting the test matrix to exclude an unsupported configuration: Python 3.13 on macOS 13. By adding an 'exclude' clause, the workflow now avoids running tests in this environment, preventing unnecessary CI failures and ensuring the pipeline only targets supported combinations of Python versions and operating systems. No other workflow logic or steps are modified.

## Changes
| File(s) | Summary |
| --- | --- |
| .github/workflows/test.yml | Added an 'exclude' clause to the test matrix to skip Python 3.13 on macOS 13, preventing tests from running on this unsupported configuration. |

## Sequence Diagram
This diagram shows the interactions between components:

```mermaid
sequenceDiagram
    title GitHub Actions Matrix Configuration Flow

    participant GH as "GitHub Actions"
    participant Runner as "Runner Selection"
    participant Matrix as "Matrix Configuration"
    participant Job as "Job Execution"
    
    GH->>Matrix: Configure matrix with Python versions and OS options
    Note over Matrix: Python: 3.9, 3.10, 3.11, 3.12, 3.13<br>OS: ubuntu-latest, macos-13, windows-latest
    
    Matrix->>Matrix: Apply exclusion rules
    Note over Matrix: New exclusion: Python 3.13 on macos-13
    
    loop For each valid matrix combination
        Matrix->>Runner: Request runner for combination
        alt OS is ubuntu-latest
            Runner-->>Job: Provide ubuntu-latest runner
        else OS is macos-13
            alt Python version is 3.13
                Runner-->>Matrix: Skip (excluded combination)
                Note over Runner,Matrix: New exclusion logic
            else Other Python versions
                Runner-->>Job: Provide macos-13 runner
            end
        else OS is windows-latest
            Runner-->>Job: Provide windows-latest runner
        end
        
        Job->>Job: Execute checkout action
        Note over Job: actions/checkout@v4
        Job->>Job: Continue with workflow steps
    end
```


▶️ ⚡ **AI Code Reviews for VS Code, Cursor, Windsurf**  
[Install the extension](https://marketplace.visualstudio.com/items?itemName=EntelligenceAI.EntelligenceAI-PR-Reviewer)

<details>
<summary>Note for Windsurf</summary>
Please change the default marketplace provider to the following in the windsurf settings:

Marketplace Extension Gallery Service URL: https://marketplace.visualstudio.com/_apis/public/gallery

Marketplace Gallery Item URL: https://marketplace.visualstudio.com/items
</details>

<details>
<summary>Entelligence.ai can learn from your feedback. Simply add 👍 / 👎 emojis to teach it your preferences. More shortcuts below</summary>

### **Emoji Descriptions:**
- ⚠️ **Potential Issue** - May require further investigation.
- 🔒 **Security Vulnerability** - Fix to ensure system safety.
- 💻 **Code Improvement** - Suggestions to enhance code quality.
- 🔨 **Refactor Suggestion** - Recommendations for restructuring code.
- ℹ️ **Others** - General comments and information.

### **Interact with the Bot:**
- Send a message or request using the format:
`@entelligenceai + *your message*`
```plaintext
Example: @entelligenceai Can you suggest improvements for this code?
```
- Help the Bot learn by providing feedback on its responses.
`@entelligenceai + *feedback*`
```plaintext
Example: @entelligenceai Do not comment on `save_auth` function !
```

Also you can trigger various commands with the bot by doing
`@entelligenceai command`

The current supported commands are
1. `config` - shows the current config
2. `retrigger_review` - retriggers the review

More commands to be added soon.
</details>

adityagarwal15 commented:
@fkiraly , i made the updated changes, can you please review it once

This pull request tackles the PyTorch installation compatibility issue that occurs specifically on the macOS 13 + Python 3.13 environment in our CI pipeline.

Previously, we had to exclude this entire environment from our test matrix because PyTorch didn’t support it yet, which led to CI failures. Instead of skipping tests altogether for this combo, this update takes a smarter approach by using environment markers in pyproject.toml. This way, PyTorch and related packages are only installed when the environment supports them.

What’s changed?

1. Updated the dependencies in pyproject.toml to include environment markers that conditionally install PyTorch and PyTorch Forecasting based on compatibility.

2. Removed the blanket exclusion of macOS 13 + Python 3.13 from the GitHub Actions test matrix.

This improvement means our CI coverage is more complete, avoiding unnecessary test skips while still preventing failures due to unsupported PyTorch versions.

Closes issue #8127

fkiraly reviewed (approved):
Nice! Let's see what happens in the tests.

adityagarwal15 commented:
> Nice! Let's see what happens in the tests.

Thanks for the review and approval! CI checks have passed. Could you please merge this when convenient?

adityagarwal15 commented:
@fkiraly  The CI workflow is stuck and needs a maintainer's approval before proceeding. 

**Issues:**
- Multiple jobs are "Waiting for pending jobs." 
- The workflow shows "requires approval from a maintainer."
- Some jobs timed out due to GitHub Actions runner shortage (24hr+ waits)

**Request:**
Please approve the pending workflow from the Actions tab and re-run any failed jobs. The environment marker approach in `pyproject.toml` should now properly handle the Python 3.13 + macOS-13 PyTorch issue.

Thanks!
